### ISSUE TITLE
Washed areas in "narrow" color spaces

### ISSUE THREAD

**Issue by teplit** [Beep6581/RawTherapee#5706]
I encountered a situation which looks abnormal. The extremely yellow areas in the image become completely white in some working color spaces.

In the wide ACESp0 color space the flowers look fine:
![Yellow flowers (ACESp0)](https://user-images.githubusercontent.com/5978882/78994897-d7b4c980-7b49-11ea-9efc-e15488ebe493.png)

In the narrower ACESp1 color space washed areas appear:
![Yellow flowers (ACESp1)](https://user-images.githubusercontent.com/5978882/78995000-12b6fd00-7b4a-11ea-83dd-4944ea382976.png)

The RAW file is taken from [PlayRaw](https://discuss.pixls.us/t/playraw-yellow-flowers-in-the-sun/10843).
The PP3 file is attached [here](https://github.com/Beep6581/RawTherapee/files/4461787/20180901_Moseltour_7108_Ernst.DNG.zip).

**Comment by Thanatomanic** [comment]
I'm pretty sure this is not directly related to the choice of color space. You have enabled Luminosity mode in RGB Curves, so things get pushed out of gamut pretty easily, especially when the original image is overexposed as it is.
Solution: turn of Luminosity mode, or set Exposure compensation to -0,50.

**Comment by teplit** [comment]
Why out of gamut colors should end as severe clipping in the three channels?
It doesn't make sense to get overexposure when you lower the luminosity of the image. With the RGB Curves disabled there is no washed areas (for ACESp1 working space).

If the behavior is not abnormal then it is not safe to use the following color spaces as a Working Profile:

- ACESp1
- Adobe RGB
- BestRGB
- Beta RGB
- BruceRGB
- Rec2020
- WideGamut

**Comment by teplit** [comment]
I decided to see what will happen with the image with even more extreme colors. So I did the following:
1. Set the Working Profile to ACESp1
2. I disabled the RGB Curves.
3. Activated L\*a\*b Adjustments and set Chromaticity to 100.

As you can see there is no washed areas in the image:
![Yellow flowers (Chromaticity)](https://user-images.githubusercontent.com/5978882/79022303-4c0d5e00-7b86-11ea-9a7f-e0f1bc2b8f13.png)

So the idea that out of gamut yellow colors can end as white areas is practically dead for me.

**Comment by teplit** [comment]
Something else - if you take a closer look of the image with washed areas you will see hard transitions between a washed area and its neighborhood. The red curve in the RGB Curves tool is however completely smooth. **This is abnormal situation.** You can disable the RGB curve tool, then copy the red curve and paste it as L\* curve in the L\*a\*b Adjustments tool to see what will happen. 

Do you see washed ares or hard transitions here?
![Yellow flowers (Lab curve)](https://user-images.githubusercontent.com/5978882/79036520-9536cf80-7bd1-11ea-8e2d-1fbc8325fda7.png)

There is something broken in the Luminosity Mode of RGB Curves tool.

**Comment by Thanatomanic** [comment]
Thanks @teplit for the investigation. The relevant code for the Luminosity mode starts here: https://github.com/Beep6581/RawTherapee/blob/1c63d75399f7a372c6f621e03cd25141e55ed5df/rtengine/improcfun.cc#L2614 Maybe @heckflosse or @Desmis can shine their light on the correctness of the code here?

However, if I apply Neutral profile and only adjust the L* curve in the Lab module, and I pull it down further than you do, I still get washed areas and hard transitions. The working color space is ProPhoto here.

![image](https://user-images.githubusercontent.com/6567747/79037742-09727280-7bd4-11ea-86ab-21c1d9ac35b3.png)

I think this behavior is perfectly normal because the image has overexposed areas. This completely destroys any useful color information in those areas, and that means that color operations can (and will!) also behave strangely. This is probably true for the Luminosity mode in the RGB Curves as well. If you're still convinced something is wrong, please try to find the same effect for an image without overexposed areas.

**Comment by teplit** [comment]
When you apply the Neutral profile this resets the Highlight compression and Highlight reconstruction. Note that Highlight recovery is activated in the attached PP3 file and it happens **before** the RGB Curves tool according to the [Toolchain Pipeline article](http://rawpedia.rawtherapee.com/Toolchain_Pipeline). Also note that you don't get blank white areas in the yellow parts of the image as illustrated in the second screenshot.

**Comment by Desmis** [comment]
Very difficult for me to express myself in English on a complex subject, but I will try.

To explain **this behavior, which seems normal to me**.

I mesure with lockable color picker, Lab values in a yellow flower, with RGB Curves disabled
AcesP0 L=96 a=3.3 b=125.4
AcesP1 L=95.8 a=3.4 b=126
Prophoto L=96 a=3.3 b=126.5
AdobeRGB L=95.1 a=3.8 b=97.1
sRGB L=93.6 a=4.5 b=87.3

If you look at R (RGB) for these values (range 0.255)
Prophoto R=250 near limit of gamut
AdobeRGB R=269 out of gamut
sRGB R=276 out of gamut
I cannot simply calculate values for ACes...but values are near Propohoto, probably low for AcesP0

Now, if we apply RGB curves, we increase hugely R channel...
AcesP0 become L=91.6 a=3.3 b=125.4 always in gamut, but color mangt. made his job and reduce L
AcesP1 become L=100 a=0 b=0 we are out of gamut, Color management is out
Prophoto become L=92.9 a=3.3 b=126.5
AdobeRGB and SRGB become L=100 a=0 b=0 we are out of gamut, Color management is out.

Now if I use Lab adjustements and set chromaticity to 58 (and disable RGB curves) we only increase C in LCH, and color management does a best job to compare to RGB mode
2 cases, without and with Avoid color shift

AcesP0 L=96 a=5.4 b=166 , with avoid Color shift (acs) L=96 a=4.1 b=154.6 (before "acs" we are out of gamut)
AcesP1 L=95.9 a=5.4 b=165.3, with acs L=95.9 a=41.1 b=158.1 (before "acs" we are out of gamut)
Prophoto L=96 a=5.4 b=165.5, with acs L=96 a=4.2 b=162.6 (before "acs" we are out of gamut)
AdobeRGB L=95.1 a=6.3 b=153.3, with acs L=96 a=4.3 b=102 (before "acs" we are out of gamut)
sRGB L=93.7 a=7.1 b=138, with acs L=93.7 a=4.8 b=91.6 (before "acs" we are out of gamut)

To note in Lab mode, and with "acs", that a Munsell correction is done to improve true color, with important corrections in blue-purple, Orange, and green

In summary it is a problem of **color management**, which goes **badly in RGB** mode, because we **control nothing**.
it mainly affects the **red channel**, because the gamut towards L = 95 is very narrow towards the reds / yellows

jacques

**Comment by teplit** [comment]
From my point of view as non-developer the behavior is **severely unsafe**, because (1) it can result in loss of visual information which manifests in this case as unwanted clipping in key parts of the image and because (2) it is not evident that it will happen.

A substantial color boost in Lab via the Chromaticity slider does not cause washed areas irrespective of the Avoid color shift option even with sRGB as Working Profile.

**Comment by Desmis** [comment]
@teplit 

I know my english is bad, but have you read my post ?

Lab mode does not create wash, because of use of Color Management
RGB mode create wash, because thers is nt color management

**Comment by teplit** [comment]
Yes, I read it, however I'm not familiar what exactly RawTherapee do with the colors in the background.

**Comment by Thanatomanic** [comment]
> When you apply the Neutral profile this resets the Highlight compression and Highlight reconstruction.

> severely unsafe, because (1) it can result in loss of visual information which manifests in this case as unwanted clipping in key parts of the image and because (2) it is not evident that it will happen.

High reconstruction tries to bring back some color information if only one or two color channels are clipped. It can only do this by making assumptions about the missing channels. In other words, as soon as you have at least one channel overexposed (saturated pixel on your sensor), your original light intensities are no longer truly recoverable, and therefore also your color is no longer truly recoverable. This means that whatever other operation you do on these recovered colors __cannot and will not behave as you would expect__. There is really no way around that.

If you find similar strange behavior with RGB Curves in properly exposed areas (without necessary recovery), then there might be a more fundamental issue.

**Comment by Thanatomanic** [comment]
> Lab mode does not create wash, because of use of Color Management
> RGB mode create wash, because thers is nt color management

@Desmis Are you sure? In the code for `RGB Levels` in `Luminosity mode` I linked before I see a conversion to the working profile, then to Lab and then there is even some gamut control going on.

**Comment by Desmis** [comment]
yes sure, because Lab is after...

If you want to do a control management with RGB, you must convert step by step in Lab mode, then call module "avoid color shift", then reconvert RGB.
When you control shift after RGB, the damage is done and almost irretrievable

I was the one who built the Lab module, with avoid color shift, and Munsell correction.

making a control on an RGB module is not impossible, but difficult

**Comment by Desmis** [comment]
it is also me who built the luminance mode in RGB

**Comment by Desmis** [comment]
but, I can look "again" and see if we can  improve... with no promises

**Comment by Desmis** [comment]
There is perhaps something wrong ?? in my "old" work ? (about 8 years ago)

**Comment by Desmis** [comment]
To be more complete, the current gamut control only works under certain conditions.
If variation of gamut are low, and if the three chanels L, a, b are not completely out.

In this case, when you increase Red, and when you convert Lab, datas are with L > 100 and a and b are quasi unknown. Gamut control in this case control nothing...

In gamut control Lab, the algorithm play principaly with C (chrominance) and sometimes with L when C condition is limit

if you want it to work in this case and more generally in all RGB mode, you have to do one (or 3) additional loops, to reduce either R, B, G, or the 3...Until L, a, and b make sense

Complex to do ! is it worth it ?

jacques

**Comment by teplit** [comment]
@Desmis Thanks for clarifying the matter.

While trying to gain better understanding of the problem I read the [Unclipped article](http://rawpedia.rawtherapee.com/Unclipped) in RawPedia. Apparently the RawTherapee engine can work with unclipped data (out of gamut colors) from input to output, provided some tools are not used (RGB Curves in Luminosity mode is explicitly mentioned).

So I did the following experiment - in ACESp1 working space I simply changed the type of the second Tone curve from Perceptual to Film-like. Voilà, this visually solved the problem - the color wash is gone.

It appears that the Perceptual curve can cause out of gamut colors which can choke subsequent tools not capable of handling unbounded colors like RGB Curves in Luminosity mode. Is this correct presumption?

**Comment by Desmis** [comment]
@teplit 
Thank you for your analyse.

I think when RGB are too high.. (and it's function of working space), all soltions are work around...

Change the algorihm RGB curves, is a big challenge, if we want it works in all cases... gamut control in RGB mode, while action on R or G or B channel is very difficult...

**Comment by Thanatomanic** [comment]
I don't think there is anything more to be done for this issue. Working with nearly or fully clipped data (as is the case in the provided example) is always going to present some problem in the pipeline. There is no concrete (theoretical) argument on how to improve the current toolset. Therefore, closing for now.

**Comment by teplit** [comment]
RawTherapee can work gracefully with RAW files containing partially or completely blown highlights provided some broken/unsafe tools are not used. There are a lot of cases where the highlight clipping is either technically unavoidable and/or artistically desirable like for example including the sun as an element in a photographed scene.

The problem is triggered by a Perceptual curve and can happen in sRGB working space. No other curve mode triggers the problem as far as I see.

**Comment by Thanatomanic** [comment]
I see your point and we should probably add something to the documentation to explain what is going on. 

There seem to be some errors or misconceptions in the documentation anyway, because from a very simple experiment, this

>"The choice of working profile has an influence on the effect of the curves in all modes **except for perceptual** - in this mode, changing the working profile will not alter the effect of the curve." (http://rawpedia.rawtherapee.com/Exposure#Perceptual)

is just not true.

**Comment by Thanatomanic** [comment]
Also, Luminosity mode in RGB Curves definitely doesn't do unbounded processing properly. Try a simple edit and then turn 'Clip out-of-gamut colors' off. That's probably very logical, but it should be documented.

**Comment by teplit** [comment]
Yes, I see a similar effect as using a Perceptual curve. This raises the question whether or not the Perceptual curve as color space independent tool can produce values outside the working color space.

**Comment by teplit** [comment]
I made an experiment to see whether or not my hypothesis for the Perceptual curve has some ground. I reproduced a similar harsh transition without the Luminosity mode – instead of it I simply inverted the RGB channels via the RGB Curves tool. Here is the effect:
![RGB inversion](https://user-images.githubusercontent.com/5978882/87870209-7b660f00-c9ae-11ea-9014-a98d0d019545.png)
[20180901_Moseltour_7108_Ernst.DNG.pp3.txt](https://github.com/Beep6581/RawTherapee/files/4943410/20180901_Moseltour_7108_Ernst.DNG.pp3.txt) (remove the .txt extension)

No other curve mode causes this harsh transition. The RGB inversion itself should not cause gamut problems at all, because in essence it is a simple mapping of the sort of (R, G, B) -> (1 − R, 1 − G, 1 − B). However if there are colors outside the working color space (Rec2020 in this case) then we have an obvious problem. Note that the _Clip out-of-gamut colors_ is enabled which should ensure that the starting point for image processing does not contain out of gamut (unbounded) colors. So I think that the Perceptual curve most likely outputs colors which fall outside the working color space. You can’t measure them with a color picker in RawTherapee 5.8.

**Comment by Beep6581** [comment]
> in essence it is a simple mapping of the sort of (R, G, B) -> (1 − R, 1 − G, 1 − B)

That's quite an assumption. Such assumptions need to be verified by looking at the code.

Some parts of RawPedia are new and correct, while others are very old and probably no longer true - either because the behavior was changed or because the behavior of the code was never that as was documented (could be due to bugs, could be due to the documentation not being written by the person who implemented the code).

The Exposure article is one such old article and probably needs to be rewritten from scratch, along with a code analysis to make sure that what is written is what the code does.

**Comment by teplit** [comment]
In Adobe Photoshop the RGB channel inversion performed via Curves adjustment is an involution as expected. If in RawTherapee this is not the case then there is another problem…
